### PR TITLE
Household composition real time changes

### PR DESCRIPTION
**Pull request by ajmaddaford**
### What is the context of this PR?

Enables the "realtime" adding/removing of household members in the household composition question.

### How to review 

1. Visit `test_repeating_household`
2. Complete household question, adding several members
3. Edit from summary screen, ensuring edited values are carried through.

### CODE REVIEW

**Comment by yoganandkunche** [comment]
QA passed on Branch. Created a new trello card following discussion with Louise M

https://trello.com/c/frP99GYd/778-census-must-not-wipe-data-if-responded-does-not-make-any-changes-on-household-who-live-here-page

File: app/data_model/answer_store.py [hunk 1]
```diff
@@ -206,16 +206,36 @@ def map(self, group_id=None, block_id=None, answer_id=None, group_instance=None,
...
+        :param answer_id:
+        :param block_id:
+        :param group_id:
+        :param answer_instance:
+        :param group_instance:
+        """
+        use_filter = (group_id or block_id or answer_id or group_instance or answer_instance) is not None
+        answers = self.filter(group_id, block_id, answer_id, group_instance, answer_instance) if use_filter else self.answers
+        for answer in answers:
+            answer_store_answer = Answer(group_id=answer['group_id'],
```
Comment (benjefferies) [inline app/data_model/answer_store.py]: Seems a bit overkill to find the answer create an answer find the answer index and then remove via index when we can just do self.answers.remove(answer)

**Review by benjefferies — CHANGES_REQUESTED** [review]

File: app/data_model/answer_store.py (hunk 1 above)
Comment (ajmaddaford) [reply app/data_model/answer_store.py]: Good point!

File: app/data_model/answer_store.py [hunk 2]
```diff
@@ -206,16 +206,32 @@ def map(self, group_id=None, block_id=None, answer_id=None, group_instance=None,
...
+        :param answer_id:
+        :param block_id:
+        :param group_id:
+        :param answer_instance:
+        :param group_instance:
+        """
+        use_filter = (group_id or block_id or answer_id or group_instance or answer_instance) is not None
+        answers = self.filter(group_id, block_id, answer_id, group_instance, answer_instance) if use_filter else self.answers
+
+        for answer in answers:
```
Comment (benjefferies) [inline app/data_model/answer_store.py]: Not sure if we can modify the list if we are iterating over it when user_filter == False. Worth a test to confirm

File: tests/app/data_model/test_answer_store.py [hunk 3]
```diff
@@ -407,5 +407,206 @@ def test_remove_answer_that_does_not_exist(self):
...
+        )
+
+        self.store.add(answer_1)
+        self.store.add(answer_2)
+
+        expected_answers = {
+            "2_1": 25,
+            "5_1": 65,
+        }
+        self.assertEqual(self.store.map(), expected_answers)
```
Comment (benjefferies) [inline tests/app/data_model/test_answer_store.py]: At this point can't we make the assumption that they have been added?

**Review by benjefferies — CHANGES_REQUESTED** [review]

File: app/data_model/answer_store.py (hunk 2 above)
Comment (ajmaddaford) [reply app/data_model/answer_store.py]: Good spot! Fixed the code and added a unit test

File: tests/app/data_model/test_answer_store.py (hunk 3 above)
Comment (ajmaddaford) [reply tests/app/data_model/test_answer_store.py]: Yeah, it was copied and pasted from another test. I've removed it from this one now.

**Comment by collisdigital** [comment]
Waiting on unit/int tests from @ajmaddaford before merging

File: tests/app/data_model/test_answer_store.py [hunk 4]
```diff
@@ -407,5 +407,220 @@ def test_remove_answer_that_does_not_exist(self):
...
+        self.store.add(answer_2)
+
+        self.store.remove(block_id="block2")
+        expected_answers = {
+            "answer1": 10,
+            "answer2": 20
+        }
+        self.assertEqual(self.store.map(), expected_answers)
+
+    def test_remove_all_answers(self):
```
Comment (benjefferies) [inline tests/app/data_model/test_answer_store.py:608]: Should we have more tests to cover each filter param?

File: tests/integration/household_composition/test_repeating_household.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
+from werkzeug.datastructures import MultiDict
+
+from tests.integration.create_token import create_token
+from tests.integration.integration_test_case import IntegrationTestCase
+
+
+class TestRepeatingHousehold(IntegrationTestCase):
```
Comment (benjefferies) [inline tests/integration/household_composition/test_repeating_household.py:7]: Definitely made sense to make these integration tests

**Review by benjefferies — APPROVED** [review]

File: tests/app/data_model/test_answer_store.py (hunk 4 above)
Comment (ajmaddaford) [reply tests/app/data_model/test_answer_store.py:608]: Possibly, my thinking was it uses filter and that has tests for each param.

File: app/templates/partials/questions/repeatinganswer.html [hunk 6]
```diff
@@ -1,22 +1,24 @@
 <div class="question question--repeatinganswer" id="{{question.id}}">
 
-  <fieldset class="question__fieldset">
+    <h2 class="question)__title neptune">{{question_title}}</h2>
```
Comment (samiwel) [inline app/templates/partials/questions/repeatinganswer.html]: Is the `)` a typo in the class name?

File: app/assets/js/app/modules/household.js [hunk 7]
```diff
@@ -15,42 +25,45 @@ class HouseholdMember {
...
 
+  add(node, parent) {
+    this.node = node
     this.removeBtn = document.createElement('button')
-    this.removeBtn.innerHTML = 'remove'
-    this.removeBtn.classList.add('btn', 'btn--link')
-    this.removeBtn.addEventListener('click', this.onRemoveClick)
 
+    this.bindToDOM(node)
+    this.removeBtn.innerHTML = 'Remove'
```
Comment (samiwel) [inline app/assets/js/app/modules/household.js]: Is this string translatable for the Welsh version of census?

File: app/data_model/answer_store.py [hunk 8]
```diff
@@ -206,16 +206,32 @@ def map(self, group_id=None, block_id=None, answer_id=None, group_instance=None,
...
+        """
+        Removes answer(s) from the answer store.
+
+        :param answer_id:
+        :param block_id:
+        :param group_id:
+        :param answer_instance:
+        :param group_instance:
+        """
+        use_filter = (group_id or block_id or answer_id or group_instance or answer_instance) is not None
```
Comment (jonnyshaw89) [inline app/data_model/answer_store.py]: could the use_filter logic potentially go into the filter function as its duplicated in a couple of places in this file. e.g. if you pass no filter arguments in then it doesn't filter?

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: app/templates/partials/questions/repeatinganswer.html (hunk 6 above)
Comment (ajmaddaford) [reply app/templates/partials/questions/repeatinganswer.html]: Good spot. Fixed.

File: app/assets/js/app/modules/household.js (hunk 7 above)
Comment (ajmaddaford) [reply app/assets/js/app/modules/household.js]: Good spot. I'll speak to @hamishtaplin tomorrow about what we can do for this

File: app/data_model/answer_store.py (hunk 8 above)
Comment (ajmaddaford) [reply app/data_model/answer_store.py]: Good point. I've moved the `use_filter` logic out of the `remove` and `map` methods. I think we should consider making the `filter` method something we can chain in the future, so we could have something like `answer_store.filter(group_id='group1').remove()`

File: app/data_model/answer_store.py [hunk 9]
```diff
@@ -154,6 +154,7 @@ def count(self, answer):
     def filter(self, group_id=None, block_id=None, answer_id=None, group_instance=None, answer_instance=None):
         """
         Find all answers in the answer store for a given set of filter parameter matches.
+        If no filter parameters are passed it returns a copy of the list of all answers.
```
Comment (jonnyshaw89) [inline app/data_model/answer_store.py:157]: Oh, so it already worked like that?

Comment (ajmaddaford) [reply app/data_model/answer_store.py:157]: Yeah

**Review by jonnyshaw89 — APPROVED** [review]

**Comment by jonnyshaw89** [comment]
👍 Lets squash this one.

**Comment by ajmaddaford** [comment]
There is still a small change to make to sort out the 'Remove' text in the Javascript

**Comment by jonnyshaw89** [comment]
@ajmaddaford Is that currently being worked on or can it be follow up?

**Comment by ajmaddaford** [comment]
@bennpearson is working on it

**Review by samiwel — APPROVED** [review]
Looking good. There'll be some work to do merging this PR with the multiple repeating answers PR though.

**Comment by ajmaddaford** [comment]
Merge waiting on #613

**Comment by benjefferies** [comment]
Once #631 is merged I can squash this one and it's ready to go?

**Comment by ajmaddaford** [comment]
@benjefferies yes

**Comment by benjefferies** [comment]
I will have to resolve all conflicts for each rebase although any merge conflicts have been resolved previously. For that reason to get the same result I will just copy the repo and create a new commit

File: tests/integration/household_composition/test_household_question.py [hunk 10]
```diff
@@ -134,37 +127,11 @@ def add_answers(self):
         self.login_and_check_introduction_text()
         first_page = self.start_questionnaire()
 
-        # Add first person
-        form_data = MultiDict()
```
Comment (benjefferies) [inline tests/integration/household_composition/test_household_question.py:138]: The add logic has been tested in another test, we don't need to test it again

**Comment by ajmaddaford** [comment]
Closing this pull request, changes raised in new pull request #696

**Pull request closed without merging**